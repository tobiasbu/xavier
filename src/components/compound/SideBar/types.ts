
export interface OptionsProps {
  children: string;
  onSelect?: null;
  selected?: boolean;
  tabIndex?: number;
}