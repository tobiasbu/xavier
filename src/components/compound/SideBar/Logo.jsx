import React from 'react';
import { createUseStyles } from 'react-jss';

import magnetosLogo from './magnetos-logo.svg';
import generateId from '../../generateId';

const useStyle = createUseStyles({
  magnetos: {
    maxWidth: '100%',
    height: 148,
    margin: 'auto',
    textIndent: '-999px',
    backgroundSize: 'contain',
    backgroundImage: `url(${magnetosLogo})`,
    backgroundRepeat: 'no-repeat',
    backgroundPosition: 'center center',
  },
  magnetosWrapper: {
    margin: '2.8rem',
  },
}, {
  generateId,
});

/**
 * Magnetos logo.
 */
const Logo = () => {
  const classes = useStyle();
  return (
    <div className={classes.magnetosWrapper}>
      <a href="/">
        <h1 className={classes.magnetos} alt="Magnetos Logo">Magnetos</h1>
      </a>
    </div>
  );
};

export default Logo;
