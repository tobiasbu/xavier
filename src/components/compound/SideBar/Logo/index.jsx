import React from 'react';

import useStyle from './style';

/**
 * Magnetos logo.
 */
const Logo = () => {
  const classes = useStyle();
  return (
    <div className={classes.magnetosWrapper}>
      <a href="/">
        <h1 className={classes.magnetos} alt="Magnetos Logo">Magnetos</h1>
      </a>
    </div>
  );
};

export default Logo;
