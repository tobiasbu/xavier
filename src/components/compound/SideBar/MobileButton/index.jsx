import React from 'react';
import PropTypes from 'prop-types';

import useStyle from './style';

const MobileButton = (props) => {
  const { onClick, open } = props;
  const classes = useStyle();
  const openClass = (open) ? classes.mobileClose : '';
  const icon = (open) ? 'close' : 'menu';
  return (
    <button
      className={`a-btn a-btn--mars a-btn--medium a-btn--icon ${classes.mobileButton} ${openClass}`}
      type="button"
      onClick={onClick}
    >
      <i className={`a-icon a-icon--${icon} a-icon--size-medium ${classes.mobileIcon}`} aria-label="Acessar menu principal" />
    </button>
  );
};

MobileButton.defaultProps = {
  open: false,
};

MobileButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  open: PropTypes.bool,
};

export default MobileButton;
