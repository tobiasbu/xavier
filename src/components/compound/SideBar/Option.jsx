import React from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';

const useStyle = createUseStyles({
  optionWrapper: {
    fontFamily: 'Poppins, sans-serif',
    fontSize: 20,
    fontWeight: 500,
    color: '#32383c',
    padding: '2rem 2.5rem',
    position: 'relative',
    display: 'block',
    textDecoration: 'none',
    transition: 'background-color 0.2s ease 0s',
    cursor: 'pointer',
    '&:hover, .selected': {
      backgroundColor: 'var(--color-moon-200)',
    },
  },
});

/**
 * Side bar option
 * @param {import('./types').OptionsProps} props Option props
 */
const Option = (props) => {
  const {
    children, onSelect, selected, tabIndex,
  } = props;

  const classes = useStyle();
  const selectedClass = (selected) ? ' selected' : '';

  return (
    <div
      className={`${classes.optionWrapper}${selectedClass}`}
      onClick={onSelect}
      onKeyDown={onSelect}
      role="button"
      tabIndex={tabIndex}
      aria-label={`${children}`}
    >
      {children}
    </div>
  );
};

Option.defaultProps = {
  onSelect: null,
  selected: false,
  tabIndex: 0,
};

Option.propTypes = {
  children: PropTypes.string.isRequired,
  onSelect: PropTypes.func,
  selected: PropTypes.bool,
  tabIndex: PropTypes.number,
};


export default Option;
