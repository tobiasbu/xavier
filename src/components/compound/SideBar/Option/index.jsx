import React from 'react';
import PropTypes from 'prop-types';

import { NavLink } from 'react-router-dom';

import useStyle from './style';

/**
 * Side bar option
 * @param {import('../types').OptionsProps} props Option props
 */
const Option = (props) => {
  const {
    children, selected, link,
  } = props;

  const classes = useStyle();
  const selectedClass = (selected) ? classes.optionSelected : '';
  const ariaCurrent = (selected) ? 'true' : 'page';

  return (
    <NavLink
      to={link}
      className={`${classes.optionWrapper} ${selectedClass}`}
      aria-label={`${children}`}
      aria-current={ariaCurrent}
    >
      {children}
    </NavLink>
  );
};

/**
 * Option default props.
 */
Option.defaultProps = {
  link: '/',
  selected: false,
};

/**
 * Option prop types
 */
Option.propTypes = {
  link: PropTypes.string,
  selected: PropTypes.bool,
  children: PropTypes.string.isRequired,
};


export default Option;
