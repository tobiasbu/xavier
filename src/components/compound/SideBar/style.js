import jss from 'jss';
import MediaQueries from '../../../style/mediaQueries';

const style = jss.createStyleSheet({
  sidebarWrapper: {
    width: 300,
    position: 'relative',
    [MediaQueries.large]: {
      width: 0,
    },
  },
  sidebarContent: {
    position: 'fixed',
    top: 0,
    left: 0,
    display: 'flex',
    flexDirection: 'column',
    width: 300,
    minWidth: 300,
    height: '100%',
    minHeight: '100vh',
    backgroundColor: 'white',
    borderRight: '1px solid var(--color-space-300)',
    transition: '0.15s ease transform',
    transform: 'translateX(0%)',
    [MediaQueries.large]: {
      transform: 'translateX(-100%)',
    },
    zIndex: 20,
  },
  sidebarOptionsContainer: {
    overflowY: 'auto',
    flex: '1 1 0%',
  },
  openContent: {
    transform: 'translateX(0%)',
  },
  sidebarMainMenu: {
    display: 'flex !important',
    flexDirection: 'column',
    position: 'fixed',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    [MediaQueries.large]: {
      display: 'none',
    },
  },
  openMainMenu: {
    display: 'flex !important',
  },
  backgroundMenu: {
    backgroundColor: 'black',
    opacity: 0.5,
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    position: 'fixed',
    zIndex: 10,
    display: 'none',
    [MediaQueries.largeUp]: {
      display: 'none !important',
    },
  },
});

export default style;
