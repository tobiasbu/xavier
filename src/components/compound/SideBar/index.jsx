import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import style from './style';
import Logo from './Logo';
import Option from './Option';
import MobileButton from './MobileButton';

import routeMap from '../../../router/routeMap.json';

/**
 * Helper function to create <Option> element.
 * @param {{ link: string, name: string }} route Route descriptor
 * @param {number} index Keyed index for React
 * @param {boolean} selected Is selected?
 */
const mapOption = (route, index, selected, onSelect) => (
  <Option
    link={route.link}
    key={index}
    selected={selected}
    onSelect={onSelect}
  >
    {route.name}
  </Option>
);

/**
 * Xavier side bar menu.
 *
 * I could totally use pure function here.
 * But here I'm testing the JSS style `attach` and  `detach` functions.
 */
class SideBar extends React.Component {
  constructor() {
    super();
    // useState can solve that.
    this.state = {
      open: false,
    };
    // Experimental arrow function in classes is disabled, so bound function:
    this.onMobileButton = this.onMobileButton.bind(this);
    this.onSelect = this.onSelect.bind(this);
  }

  componentDidMount() {
    style.attach();
  }

  componentWillUnmount() {
    style.detach();
  }


  onMobileButton(e) {
    e.preventDefault();
    this.triggerMenu();
  }

  onSelect() {
    this.triggerMenu();
  }

  triggerMenu() {
    const { open } = this.state;
    const negate = !open;
    this.setState({ open: negate });
  }

  render() {
    const { open } = this.state;
    const { classes } = style;
    const { location } = this.props;

    const openMainMenu = open ? classes.openMainMenu : '';

    return (
      <>
        <MobileButton onClick={this.onMobileButton} open={open} />
        <div className={classes.sidebarWrapper}>
          <div className={`${classes.sidebarContent} ${open ? classes.openContent : ''}`}>
            <div className={`${classes.sidebarMainMenu} ${openMainMenu}`}>
              <Logo />
              {routeMap.map((value, index) => {
                const isSelected = location.pathname === value.link;
                return mapOption(value, index, isSelected, this.onSelect);
              })}
            </div>
          </div>
        </div>
        <div className={`${classes.backgroundMenu} ${openMainMenu}`} />
      </>
    );
  }
}

/**
 * Side bar props
 */
SideBar.defaultProps = {
  location: {
    pathname: '',
  },
};

/**
 * Side bar prop types
 */
SideBar.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }),
};

export default withRouter(SideBar);
