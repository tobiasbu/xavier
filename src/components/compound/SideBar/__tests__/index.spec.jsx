import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import SideBar from '..';

describe('SideBar', () => {
  it('should render', () => {
    const component = shallow(<SideBar />);
    expect(toJSON(component)).toMatchSnapshot();
  });
});
