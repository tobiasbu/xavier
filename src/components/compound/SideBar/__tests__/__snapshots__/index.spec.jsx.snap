// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar should render 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
